### PR TITLE
perf(executor): improving Sentencizer by a factor of 2.5

### PR DESCRIPTION
**Pull request by guiferviz**
The Sententicizer is now 2.5 times faster.
Instead of replacing separators by a '\n' character and splitting later by the new line character, this improved version does that with one regex and using the `findall` method.
The complexity of the new algorithm is also linear but with a low constant time because it uses only one pass over the input string.
The regex also ignores spaces at the beginning and end of the sentences so no `strip` needed.

Adding two tests that show examples of questionable results of the Sentencizer.
If that is the expected behaviour, then it is good to keep those tests here to show that it is not considered a bug.
Those test are:
* Splitting by separators that appear in common constructions like floating point numbers.
* Splitting by the '\n' character. In this efficient version I've included '\n' among the separators to replicate the same behaviour as the previous version, but if '\n' it's not included among the separators then this new version is not going to split by new lines. If this behaviour is necessary, it can easily be modified. If it is not necessary, then it could be removed directly from the separators. Let me know and I will update the code.

## Time improvement

I've test the performance using an stress test that I've pull requested to the stress test repository:
https://github.com/jina-ai/stress-test/pull/1

Hardware details:
* Processor: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
* Memory: 16GiB SODIMM DDR4 Synchronous 2667 MHz (0,4 ns)
* OS: Linux Mint 19.3

Old code times:

    41.94MB read 3 times.
    Average per iteration: 2.876222

New code:

    41.94MB read 3 times                                                                                                       
    Average per iteration: 1.151864

Performance improvement: 2.4970
The results of both versions are exactly the same in the tested text.

### CODE REVIEW

**Comment by hanxiao** [comment]
🤯 geez man! great work!

**Comment by hanxiao** [comment]
@guiferviz thanks for your contribution, could you please resolve the conflict first so that the CICD could start. Btw, `self.min_sent_len < self.max_sent_len` is somehow still in your PR?

File: jina/executors/crafters/nlp/split.py [hunk 1]
```diff
@@ -31,28 +31,33 @@ def __init__(self,
...
-                                '！', '．', '？', '｡', '。']
+                                '！', '．', '？', '｡', '。', '\n']
         if self.min_sent_len < self.max_sent_len:
             self.logger.warning('the min_sent_len (={}) should be smaller or equal to the max_sent_len (={})'.format(
                 self.min_sent_len, self.max_sent_len))
-        self._slit_pat = re.compile('([{0}])+([^{0}])'.format(''.join(self.punct_chars)))
+        self._slit_pat = re.compile('\s*([^{0}]+)(?<!\s)[{0}]*'.format(''.join(self.punct_chars)))
 
-    def craft(self, raw_bytes: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
+    def craft(self, raw_bytes: bytes, doc_id: int,
```
Comment (nan-wang) [inline jina/executors/crafters/nlp/split.py:40]: please reformat this line.

**Comment by nan-wang** [comment]
As you have seen, `Sentencizer` is extremely simply for the time being. If the users want a customized rules, it shouldn't be very difficult to implement their own version based on regex or even use the punkt tokenizer from their favorite toolkits (e.g. `nltk`, `spaCy`, and etc.). For our side, we want to keep the core part of `jina` as lightweight as possible, and suggest the users to develop their own executors and share with the community in the `jina-hub`.

**Comment by guiferviz** [comment]
I just fixed the merge conflicts. Sorry about not do in yesterday.

@hanxiao `self.min_sent_len < self.max_sent_len` is included now. When I started the branch my previous pull request had not yet been accepted, so the change was not there.

@nan-wang Ok, I understand. So I think the Sentencizer I am suggesting is simple and fast, it meets your idea of what a Sentencizer should be. As you say there are very good tokenizers out there that users can use if they want something more personalized.

**Comment by guiferviz** [comment]
I'm getting the following error in one commit message.

Message:

    test(executor): Sentencizer trimming spaces

Error:

    subject must not be sentence-case, start-case, pascal-case, upper-case [subject-case]

Should I do something about this? It was in upper-case because it's the name of the class... :worried:

**Comment by hanxiao** [comment]
yes, that's commit lint. please use `git commit --amend` change your commit message and then `git push --force`

File: jina/executors/crafters/nlp/split.py (hunk 1 above)
Comment (guiferviz) [reply jina/executors/crafters/nlp/split.py]: Done! I have a habit of breaking all lines longer than 80 characters :grimacing:

**Comment by hanxiao** [comment]
recheckcla

**Pull request merged**